### ISSUE TITLE
gitlab: gitaly throws warnings at the user

### ISSUE THREAD

**Issue by petabyteboy** [NixOS/nixpkgs#79374]
**Describe the bug**
When pushing to my gitlab instance it throws warnings at the user

**To Reproduce**
1. Set up a gitlab instance
2. interact with gitaly (push, fetch)
3. see the warnings, hopefully

**Expected behavior**
No warnings should be shown

**Screenshots**

```
remote: /nix/store/j8j1nmr8k4lh03h5sh4zb8wf4skjcjs9-ruby2.6.5-json-2.2.0/lib/ruby/gems/2.6.0/gems/json-2.2.0/lib/json/common.rb:100: warning: already initialized constant JSON::NaN
remote: /nix/store/7za614zraf7jzby6zahz56qgs9kmchgs-gitaly-env/lib/ruby/gems/2.6.0/gems/json-2.2.0/lib/json/common.rb:100: warning: previous definition of NaN was here
remote: /nix/store/j8j1nmr8k4lh03h5sh4zb8wf4skjcjs9-ruby2.6.5-json-2.2.0/lib/ruby/gems/2.6.0/gems/json-2.2.0/lib/json/common.rb:102: warning: already initialized constant JSON::Infinity
remote: /nix/store/7za614zraf7jzby6zahz56qgs9kmchgs-gitaly-env/lib/ruby/gems/2.6.0/gems/json-2.2.0/lib/json/common.rb:102: warning: previous definition of Infinity was here
remote: /nix/store/j8j1nmr8k4lh03h5sh4zb8wf4skjcjs9-ruby2.6.5-json-2.2.0/lib/ruby/gems/2.6.0/gems/json-2.2.0/lib/json/common.rb:104: warning: already initialized constant JSON::MinusInfinity
remote: /nix/store/7za614zraf7jzby6zahz56qgs9kmchgs-gitaly-env/lib/ruby/gems/2.6.0/gems/json-2.2.0/lib/json/common.rb:104: warning: previous definition of MinusInfinity was here
remote: /nix/store/j8j1nmr8k4lh03h5sh4zb8wf4skjcjs9-ruby2.6.5-json-2.2.0/lib/ruby/gems/2.6.0/gems/json-2.2.0/lib/json/common.rb:129: warning: already initialized constant JSON::UnparserError
remote: /nix/store/7za614zraf7jzby6zahz56qgs9kmchgs-gitaly-env/lib/ruby/gems/2.6.0/gems/json-2.2.0/lib/json/common.rb:129: warning: previous definition of UnparserError was here
remote: /nix/store/j8j1nmr8k4lh03h5sh4zb8wf4skjcjs9-ruby2.6.5-json-2.2.0/lib/ruby/gems/2.6.0/gems/json-2.2.0/lib/json/common.rb:100: warning: already initialized constant JSON::NaN
remote: /nix/store/7za614zraf7jzby6zahz56qgs9kmchgs-gitaly-env/lib/ruby/gems/2.6.0/gems/json-2.2.0/lib/json/common.rb:100: warning: previous definition of NaN was here
remote: /nix/store/j8j1nmr8k4lh03h5sh4zb8wf4skjcjs9-ruby2.6.5-json-2.2.0/lib/ruby/gems/2.6.0/gems/json-2.2.0/lib/json/common.rb:102: warning: already initialized constant JSON::Infinity
remote: /nix/store/7za614zraf7jzby6zahz56qgs9kmchgs-gitaly-env/lib/ruby/gems/2.6.0/gems/json-2.2.0/lib/json/common.rb:102: warning: previous definition of Infinity was here
remote: /nix/store/j8j1nmr8k4lh03h5sh4zb8wf4skjcjs9-ruby2.6.5-json-2.2.0/lib/ruby/gems/2.6.0/gems/json-2.2.0/lib/json/common.rb:104: warning: already initialized constant JSON::MinusInfinity
remote: /nix/store/7za614zraf7jzby6zahz56qgs9kmchgs-gitaly-env/lib/ruby/gems/2.6.0/gems/json-2.2.0/lib/json/common.rb:104: warning: previous definition of MinusInfinity was here
remote: /nix/store/j8j1nmr8k4lh03h5sh4zb8wf4skjcjs9-ruby2.6.5-json-2.2.0/lib/ruby/gems/2.6.0/gems/json-2.2.0/lib/json/common.rb:129: warning: already initialized constant JSON::UnparserError
remote: /nix/store/7za614zraf7jzby6zahz56qgs9kmchgs-gitaly-env/lib/ruby/gems/2.6.0/gems/json-2.2.0/lib/json/common.rb:129: warning: previous definition of UnparserError was here
remote: /nix/store/j8j1nmr8k4lh03h5sh4zb8wf4skjcjs9-ruby2.6.5-json-2.2.0/lib/ruby/gems/2.6.0/gems/json-2.2.0/lib/json/common.rb:100: warning: already initialized constant JSON::NaN
remote: /nix/store/7za614zraf7jzby6zahz56qgs9kmchgs-gitaly-env/lib/ruby/gems/2.6.0/gems/json-2.2.0/lib/json/common.rb:100: warning: previous definition of NaN was here
remote: /nix/store/j8j1nmr8k4lh03h5sh4zb8wf4skjcjs9-ruby2.6.5-json-2.2.0/lib/ruby/gems/2.6.0/gems/json-2.2.0/lib/json/common.rb:102: warning: already initialized constant JSON::Infinity
remote: /nix/store/7za614zraf7jzby6zahz56qgs9kmchgs-gitaly-env/lib/ruby/gems/2.6.0/gems/json-2.2.0/lib/json/common.rb:102: warning: previous definition of Infinity was here
remote: /nix/store/j8j1nmr8k4lh03h5sh4zb8wf4skjcjs9-ruby2.6.5-json-2.2.0/lib/ruby/gems/2.6.0/gems/json-2.2.0/lib/json/common.rb:104: warning: already initialized constant JSON::MinusInfinity
remote: /nix/store/7za614zraf7jzby6zahz56qgs9kmchgs-gitaly-env/lib/ruby/gems/2.6.0/gems/json-2.2.0/lib/json/common.rb:104: warning: previous definition of MinusInfinity was here
remote: /nix/store/j8j1nmr8k4lh03h5sh4zb8wf4skjcjs9-ruby2.6.5-json-2.2.0/lib/ruby/gems/2.6.0/gems/json-2.2.0/lib/json/common.rb:129: warning: already initialized constant JSON::UnparserError
remote: /nix/store/7za614zraf7jzby6zahz56qgs9kmchgs-gitaly-env/lib/ruby/gems/2.6.0/gems/json-2.2.0/lib/json/common.rb:129: warning: previous definition of UnparserError was here
```

**Metadata**
 - system: `"x86_64-linux"`
 - host os: `Linux 5.5.0, NixOS, 20.03pre-git (Markhor)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.2`
 - channels(mil): `""`
 - nixpkgs: `/var/src/nixpkgs`

(nixpkgs is at a21c2fa3ea2b88e698db6fc151d9c7259ae14d96)

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute: gitaly
# a list of nixos modules affected by the problem
module: gitlab
```

**Comment by petabyteboy** [comment]
CC maintainers: @fpletz @globin @krav @talyz

**Comment by jslight90** [comment]
I am experiencing this same warning message with nixpkgs pinned to 71be729a58e8e4acdc173a2e7b263402eaa9b117.  It looks like the json ruby gem is getting loaded from two separate paths.  One path is just a symlink to the other.

```
remote: /nix/store/m36nfridl57qzzz5jsjrb1d5lxww5255-ruby2.6.5-json-2.2.0/lib/ruby/gems/2.6.0/gems/json-2.2.0/lib/json/common.rb:100: warning: already initialized constant JSON::NaN
remote: /nix/store/5h4i43yk9nakdf4wn2j4yp79sfm2f8gg-gitaly-env/lib/ruby/gems/2.6.0/gems/json-2.2.0/lib/json/common.rb:100: warning: previous definition of NaN was here
remote: /nix/store/m36nfridl57qzzz5jsjrb1d5lxww5255-ruby2.6.5-json-2.2.0/lib/ruby/gems/2.6.0/gems/json-2.2.0/lib/json/common.rb:102: warning: already initialized constant JSON::Infinity
remote: /nix/store/5h4i43yk9nakdf4wn2j4yp79sfm2f8gg-gitaly-env/lib/ruby/gems/2.6.0/gems/json-2.2.0/lib/json/common.rb:102: warning: previous definition of Infinity was here
remote: /nix/store/m36nfridl57qzzz5jsjrb1d5lxww5255-ruby2.6.5-json-2.2.0/lib/ruby/gems/2.6.0/gems/json-2.2.0/lib/json/common.rb:104: warning: already initialized constant JSON::MinusInfinity
remote: /nix/store/5h4i43yk9nakdf4wn2j4yp79sfm2f8gg-gitaly-env/lib/ruby/gems/2.6.0/gems/json-2.2.0/lib/json/common.rb:104: warning: previous definition of MinusInfinity was here
remote: /nix/store/m36nfridl57qzzz5jsjrb1d5lxww5255-ruby2.6.5-json-2.2.0/lib/ruby/gems/2.6.0/gems/json-2.2.0/lib/json/common.rb:129: warning: already initialized constant JSON::UnparserError
remote: /nix/store/5h4i43yk9nakdf4wn2j4yp79sfm2f8gg-gitaly-env/lib/ruby/gems/2.6.0/gems/json-2.2.0/lib/json/common.rb:129: warning: previous definition of UnparserError was here
remote: /nix/store/m36nfridl57qzzz5jsjrb1d5lxww5255-ruby2.6.5-json-2.2.0/lib/ruby/gems/2.6.0/gems/json-2.2.0/lib/json/common.rb:100: warning: already initialized constant JSON::NaN
remote: /nix/store/5h4i43yk9nakdf4wn2j4yp79sfm2f8gg-gitaly-env/lib/ruby/gems/2.6.0/gems/json-2.2.0/lib/json/common.rb:100: warning: previous definition of NaN was here
remote: /nix/store/m36nfridl57qzzz5jsjrb1d5lxww5255-ruby2.6.5-json-2.2.0/lib/ruby/gems/2.6.0/gems/json-2.2.0/lib/json/common.rb:102: warning: already initialized constant JSON::Infinity
remote: /nix/store/5h4i43yk9nakdf4wn2j4yp79sfm2f8gg-gitaly-env/lib/ruby/gems/2.6.0/gems/json-2.2.0/lib/json/common.rb:102: warning: previous definition of Infinity was here
remote: /nix/store/m36nfridl57qzzz5jsjrb1d5lxww5255-ruby2.6.5-json-2.2.0/lib/ruby/gems/2.6.0/gems/json-2.2.0/lib/json/common.rb:104: warning: already initialized constant JSON::MinusInfinity
remote: /nix/store/5h4i43yk9nakdf4wn2j4yp79sfm2f8gg-gitaly-env/lib/ruby/gems/2.6.0/gems/json-2.2.0/lib/json/common.rb:104: warning: previous definition of MinusInfinity was here
remote: /nix/store/m36nfridl57qzzz5jsjrb1d5lxww5255-ruby2.6.5-json-2.2.0/lib/ruby/gems/2.6.0/gems/json-2.2.0/lib/json/common.rb:129: warning: already initialized constant JSON::UnparserError
remote: /nix/store/5h4i43yk9nakdf4wn2j4yp79sfm2f8gg-gitaly-env/lib/ruby/gems/2.6.0/gems/json-2.2.0/lib/json/common.rb:129: warning: previous definition of UnparserError was here
remote: /nix/store/m36nfridl57qzzz5jsjrb1d5lxww5255-ruby2.6.5-json-2.2.0/lib/ruby/gems/2.6.0/gems/json-2.2.0/lib/json/common.rb:100: warning: already initialized constant JSON::NaN
remote: /nix/store/5h4i43yk9nakdf4wn2j4yp79sfm2f8gg-gitaly-env/lib/ruby/gems/2.6.0/gems/json-2.2.0/lib/json/common.rb:100: warning: previous definition of NaN was here
remote: /nix/store/m36nfridl57qzzz5jsjrb1d5lxww5255-ruby2.6.5-json-2.2.0/lib/ruby/gems/2.6.0/gems/json-2.2.0/lib/json/common.rb:102: warning: already initialized constant JSON::Infinity
remote: /nix/store/5h4i43yk9nakdf4wn2j4yp79sfm2f8gg-gitaly-env/lib/ruby/gems/2.6.0/gems/json-2.2.0/lib/json/common.rb:102: warning: previous definition of Infinity was here
remote: /nix/store/m36nfridl57qzzz5jsjrb1d5lxww5255-ruby2.6.5-json-2.2.0/lib/ruby/gems/2.6.0/gems/json-2.2.0/lib/json/common.rb:104: warning: already initialized constant JSON::MinusInfinity
remote: /nix/store/5h4i43yk9nakdf4wn2j4yp79sfm2f8gg-gitaly-env/lib/ruby/gems/2.6.0/gems/json-2.2.0/lib/json/common.rb:104: warning: previous definition of MinusInfinity was here
remote: /nix/store/m36nfridl57qzzz5jsjrb1d5lxww5255-ruby2.6.5-json-2.2.0/lib/ruby/gems/2.6.0/gems/json-2.2.0/lib/json/common.rb:129: warning: already initialized constant JSON::UnparserError
remote: /nix/store/5h4i43yk9nakdf4wn2j4yp79sfm2f8gg-gitaly-env/lib/ruby/gems/2.6.0/gems/json-2.2.0/lib/json/common.rb:129: warning: previous definition of UnparserError was here
```

**Comment by mawis** [comment]
I can confirm the problem, it occurs on my GitLab server as well since I changed to NixOS 20.03pre.

**Comment by talyz** [comment]
Sorry for replying so late to this - I tried investigating this [a while ago](https://github.com/NixOS/nixpkgs/pull/77645#issuecomment-575194433), but it still puzzles me. I get why it reports it and printing the variable which tracks required paths shows that both paths, the symlink and the real path, are indeed loaded, but why? The good news is that this seems totally benign, just irritating...

**Comment by petabyteboy** [comment]
Fixed with #81556
